File: api/upload-image.js
const tencentcloud = require("tencentcloud-sdk-nodejs-lke");
const COS = require('cos-nodejs-sdk-v5');
const multiparty = require('multiparty');
const fs = require('fs');

// 配置常量
const REGION = "ap-guangzhou";
const TYPE_KEY_REALTIME = "realtime";

// 从环境变量读取配置
const SECRET_ID = process.env.SECRET_ID;
const SECRET_KEY = process.env.SECRET_KEY;
const BOT_BIZ_ID = process.env.BOT_BIZ_ID;

// 获取临时密钥
async function getTemporaryCredentials(botBizId, fileType, isPublic, typeKey) {
    const client = new tencentcloud.lke.v20231130.Client({
        credential: {
            secretId: SECRET_ID,
            secretKey: SECRET_KEY,
        },
        region: REGION,
    });

    const params = {
        BotBizId: botBizId,
        FileType: fileType,
        TypeKey: typeKey,
        IsPublic: isPublic
    };

    const response = await client.DescribeStorageCredential(params);
    const credentials = response.Credentials;

    return {
        TmpSecretId: credentials.TmpSecretId,
        TmpSecretKey: credentials.TmpSecretKey,
        Token: credentials.Token,
        UploadPath: response.UploadPath,
        Bucket: response.Bucket,
        Region: response.Region,
        Type: response.Type
    };
}

// 上传文件到 COS
async function uploadFileToCOS(fileBuffer, credentials) {
    const cos = new COS({
        SecretId: credentials.TmpSecretId,
        SecretKey: credentials.TmpSecretKey,
        SecurityToken: credentials.Token,
    });

    return new Promise((resolve, reject) => {
        cos.putObject({
            Bucket: credentials.Bucket,
            Region: credentials.Region,
            Key: credentials.UploadPath,
            Body: fileBuffer,
        }, (err, data) => {
            if (err) {
                reject(err);
            } else {
                const bucketUrl = `https://${credentials.Bucket}.${credentials.Type}.${credentials.Region}.myqcloud.com`;
                const cosUrl = `${bucketUrl}${credentials.UploadPath}`;
                resolve({ url: cosUrl });
            }
        });
    });
}

module.exports = async (req, res) => {
    // 处理 CORS
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', '*');

    if (req.method === 'OPTIONS') {
        return res.status(200).end();
    }

    if (req.method !== 'POST') {
        return res.status(405).json({ error: 'Method not allowed' });
    }

    try {
        // 解析 multipart/form-data
        const form = new multiparty.Form();

        form.parse(req, async (err, fields, files) => {
            if (err) {
                console.error('Parse error:', err);
                return res.status(500).json({ error: err.message });
            }

            const fileArray = files.file;
            if (!fileArray || fileArray.length === 0) {
                return res.status(400).json({ error: '没有文件' });
            }

            const file = fileArray[0];
            const fileBuffer = fs.readFileSync(file.path);
            const fileName = file.originalFilename;
            
            // 获取文件扩展名（不带点号）
            let fileExt = fileName.substring(fileName.lastIndexOf('.') + 1).toLowerCase();
            
            // 支持的图片格式
            const supportedFormats = ['jpg', 'jpeg', 'png', 'gif', 'bmp', 'webp'];
            if (!supportedFormats.includes(fileExt)) {
                fileExt = 'jpg'; // 默认使用 jpg
            }
            
            console.log('📁 文件信息:', {
                originalName: fileName,
                extension: fileExt,
                size: `${(fileBuffer.length / 1024 / 1024).toFixed(2)}MB`
            });

            try {
                // 获取临时密钥
                const credentials = await getTemporaryCredentials(
                    BOT_BIZ_ID,
                    fileExt,
                    true,
                    TYPE_KEY_REALTIME
                );

                // 上传到 COS
                const result = await uploadFileToCOS(fileBuffer, credentials);

                // 清理临时文件
                fs.unlinkSync(file.path);

                return res.status(200).json({
                    success: true,
                    url: result.url
                });
            } catch (uploadError) {
                console.error('Upload error:', uploadError);
                return res.status(500).json({ error: uploadError.message });
            }
        });
    } catch (error) {
        console.error('Handler error:', error);
        return res.status(500).json({ error: error.message });
    }
};

File: api/upload-image.py
from http.server import BaseHTTPRequestHandler
import json
import os
import cgi
from io import BytesIO
from qcloud_cos import CosConfig, CosS3Client
from tencentcloud.common import credential
from tencentcloud.common.common_client import CommonClient
from tencentcloud.common.profile.client_profile import ClientProfile
from tencentcloud.common.profile.http_profile import HttpProfile

# 配置常量
REGION = "ap-guangzhou"
ENDPOINT = "lke.tencentcloudapi.com"
TYPE_KEY_REALTIME = "realtime"

# 从环境变量读取配置
SECRET_ID = os.getenv('SECRET_ID')
SECRET_KEY = os.getenv('SECRET_KEY')
BOT_BIZ_ID = os.getenv('BOT_BIZ_ID', '')

def get_temporary_credentials(bot_biz_id: str, file_type: str, is_public: bool, type_key: str) -> dict:
    """获取腾讯云 COS 临时密钥"""
    try:
        cred = credential.Credential(SECRET_ID, SECRET_KEY)
        http_profile = HttpProfile()
        http_profile.endpoint = ENDPOINT
        client_profile = ClientProfile()
        client_profile.httpProfile = http_profile

        params = {
            "BotBizId": bot_biz_id,
            "FileType": file_type,
            "TypeKey": type_key,
            "IsPublic": is_public
        }
        
        common_client = CommonClient("lke", "2023-11-30", cred, REGION, profile=client_profile)
        response = common_client.call_json("DescribeStorageCredential", params)
        
        credentials = response['Response']['Credentials']
        upload_path = response['Response']['UploadPath']
        bucket = response['Response']['Bucket']
        region = response['Response']['Region']
        cos_type = response['Response']['Type']

        return {
            "TmpSecretId": credentials['TmpSecretId'],
            "TmpSecretKey": credentials['TmpSecretKey'],
            "Token": credentials['Token'],
            "UploadPath": upload_path,
            "Bucket": bucket,
            "Region": region,
            "Type": cos_type
        }
    except Exception as err:
        raise Exception(f"获取临时密钥失败: {err}")

def upload_file_to_cos(file_data: bytes, file_name: str, credentials: dict) -> dict:
    """将文件上传到腾讯云 COS"""
    config = CosConfig(
        Region=credentials['Region'],
        SecretId=credentials['TmpSecretId'],
        SecretKey=credentials['TmpSecretKey'],
        Token=credentials['Token'],
        Scheme='https'
    )
    client = CosS3Client(config)

    # 直接上传字节数据
    response = client.put_object(
        Bucket=credentials['Bucket'],
        Key=credentials['UploadPath'],
        Body=file_data,
        EnableMD5=False
    )

    bucket_url = f"https://{credentials['Bucket']}.{credentials['Type']}.{credentials['Region']}.myqcloud.com"
    cos_final_url = f"{bucket_url}{credentials['UploadPath']}"
    
    return {
        "url": cos_final_url,
        "cosPath": credentials['UploadPath'],
        "fileName": file_name
    }

class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', '*')
        self.end_headers()

    def do_POST(self):
        try:
            # 解析 multipart/form-data
            content_type = self.headers['Content-Type']
            if 'multipart/form-data' not in content_type:
                raise Exception('Invalid content type')
            
            # 获取文件数据
            form = cgi.FieldStorage(
                fp=self.rfile,
                headers=self.headers,
                environ={'REQUEST_METHOD': 'POST'}
            )
            
            if 'file' not in form:
                raise Exception('No file uploaded')
            
            file_item = form['file']
            file_data = file_item.file.read()
            file_name = file_item.filename
            file_ext = os.path.splitext(file_name)[1].lower()
            
            # 获取临时密钥
            credentials = get_temporary_credentials(
                bot_biz_id=BOT_BIZ_ID,
                file_type=file_ext,
                is_public=True,
                type_key=TYPE_KEY_REALTIME
            )
            
            # 上传到 COS
            result = upload_file_to_cos(file_data, file_name, credentials)
            
            # 返回成功响应
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            
            response_data = {
                'success': True,
                'imageUrl': result['url'],
                'cosPath': result['cosPath']
            }
            self.wfile.write(json.dumps(response_data).encode())
            
        except Exception as e:
            self.send_response(500)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            
            error_response = {'error': str(e)}
            self.wfile.write(json.dumps(error_response).encode())
